proposals/CloudIdentifier.md

```diff
@@ -0,0 +1,197 @@
+
+With the proposed changes, the web application could check whether the given file handle is already synced by a cloud storage and generate identifiers for these without a duplicate upload.
+
+#### Drag & Drop into Mail
+
+Sharing files via mail used to be done using simple file attachments. Especially for large files, it is preferable to instead upload them to a CSP and then share an access link to that file to the recipient. Google Mail’s web application, for example, already does this by prompting the user to upload a file to Google Drive when attaching a large file.
+
+With the proposed changes, the web application could already detect that the file is synced by cloud storage and generate a share link for that file’s cloud identifier without requiring the user to manually go through these steps or prompt the user to upload the file to cloud storage again.
+
+### Non-Goals
+
+This proposal does not plan to provide a standardized way for CSP’s web APIs to 
+* give access permissions to these files/directories.
+* interact (fetch/modify/etc) with these files/directories.
+* provide additional meta-data (e.g. sync status)
+
+Instead, it should only generate an opaque identifier for a file/directory so that web apps can then interact with these individual CSPs’ web APIs on the same file/directory. Obtaining the required access permissions on that file is delegated to the web app.
+Web apps might only support one specific CSP. When supporting multiple CSPs, each CSP will require their own implementation, although semantically they would likely perform similar steps.
+
+## Design
+
+![image](./images/cloud-identifier/design.png "Design")
+
+1. Locally installed CSP client registers itself as a provider for certain directories with the browser
+2. Web app receives `FileSystemHandle` for a file/directory (via File Handling opening flow, File System Access file/directory picker or drag & drop) and requests its cloud identifier(s) via `getCloudIdentifiers()`.
+3. Browser checks whether this path has been registered as being synced by a provider
+   * If no, resolve promise with empty list
+4. For each registered provider for that path, the browser will send a request to the registered provider’s executable with the file’s/directory’s path
+5. Provider requests a token for that path from the CSP
+   * The token can also be a stable and already cached identifier, in which case step 5 and 6 are skipped
+6.  CSP can choose to generate a one-time token or stable identifier to later be used by the CSP’s web APIs. Either way, the properties of the token are up to the CSP and completely opaque to the browser.
+7. Provider responds to browser with the token
```

Comment:
Meta comment: this explainer might be easier to follow if the terminology was more consistent. For example, "CSP" and "provider" are mostly used interchangeably, though for someone not intimately familiar with the problem space it might not be clear that these refer to the same thing (not to mention that a "CSP" itself has multiple parts - the local sync client and the server)

Reply:
Done, should be more clear in regards to "local sync client" or "CSP backend" now